### ISSUE TITLE
Metrics to track

### ISSUE THREAD

**Issue by duckinator** [duckinator/how_is#1]
Metrics I want to gather:

Things to include in each report:

* repository name
  * [x] JSON
  * [x] PDF
* number of open issues
  * [x] JSON
  * [x] PDF
* number of open PRs
  * [x] JSON
  * [x] PDF
* number of issues associated with each label + no label
  * [x] JSON
  * [x] PDF
* number of issues with no label
  * [x] JSON
  * [x] PDF
* average issue age
  * [x] JSON
  * [x] PDF
* average PR age
  * [x] JSON
  * [x] PDF
* date oldest issue was opened
  * [x] JSON
  * [x] PDF
* date oldest pull request was opened
  * [x] JSON
  * [x] PDF

Metrics that track both issues and PRs track them separately. E.g., one metric for total number of issues and a separate one for total number of PRs.

Tasks will be checked off when they show up in the final resulting file (which doesn't currently exist at all).

**Comment by LynnCo** [comment]
Does github's API include whether or not a commenter on an issue is a
member of the team? I'm really interested in measuring response time on
issues from team members

**Comment by duckinator** [comment]
@LynnCo I don't think so, but I can get a list of users who are a member of the team and compare manually: https://api.github.com/orgs/rubygems/members

I'm not sure if using authentication would let that list include private members or not.

**Comment by duckinator** [comment]
@LynnCo I'm also not sure how to make it a useful metric. Averaging it all would pull in information from when rubygems was very poorly maintained. That can be resolved by setting a cutoff date, but how would you decide on a cutoff date if you? Maybe all-time, past 6 months, and past month?

**Comment by duckinator** [comment]
@indirect ^ thoughts? Anything I missed that you think I should try to track? Feel free to @-mention anyone who may have ideas.

**Comment by LynnCo** [comment]
I'm interested in the week to week response time, and making a graph out go
that

**Comment by LynnCo** [comment]
Also, because I know how I work, I want a count / % of issues that have no
labels on them

**Comment by duckinator** [comment]
Also maybe count / % of issues/PRs with a particular label?

Interesting ones for RubyGems would be:

* triage
* administrative
* blocked
* major bump

**Comment by duckinator** [comment]
I'm gonna go to bed. Feel free to drop ideas/feedback here or @-mention anyone who may have ideas.

cc @indirect

**Comment by LynnCo** [comment]
Yea that's what I'm looking for. I'm going to change either RubyGems or
Bundler (probably both) so they have the same issue semantics,  and
tracking the count / % of issues with no label or triage is my main
interest

**Comment by duckinator** [comment]
Updated issue to represent my plans given this thread.

**Comment by duckinator** [comment]
~1/2 of CSV export implemented, a few things implemented for PDF. The PDF looks really, really rough atm lol

**Comment by duckinator** [comment]
~1/2 of both, now

**Comment by duckinator** [comment]
![screenshot 20](https://cloud.githubusercontent.com/assets/39698/15795545/94427dd4-29c0-11e6-84d7-d5988c585898.png)

**Comment by duckinator** [comment]
![screenshot 21](https://cloud.githubusercontent.com/assets/39698/15796672/14254b62-29d0-11e6-99de-5a76ca965c01.png)

**Comment by duckinator** [comment]
Apparently GitHub's API doesn't include labels for PRs, but does for issues?! Come on, GItHub.

**Comment by duckinator** [comment]
Okay, clearly just dumping this isn't working.
![screenshot 22](https://cloud.githubusercontent.com/assets/39698/15797110/84e33be2-29db-11e6-931e-06a8d9b65e71.png)
![screenshot 23](https://cloud.githubusercontent.com/assets/39698/15797111/84e5cbe6-29db-11e6-8aad-3365b9874276.png)

**Comment by LynnCo** [comment]
So labels on PRs is actually something I wanted to bring up about RubyGems. I am actually against labeling PRs, primarily because PRs are usually attached to issues and I expect the attached issues to contain the relevant labels. For this reason, labeling PRs tends to produce a lot of duplicate work. This is exemplified by RubyGems having [bug report](https://github.com/rubygems/rubygems/issues?utf8=%E2%9C%93&q=label%3A%22bug+report%22+) and [bug fix](https://github.com/rubygems/rubygems/issues?utf8=%E2%9C%93&q=label%3A%22bug+fix%22+). Github not including PR labels in their API makes me think that they agree with me on this.

**Comment by LynnCo** [comment]
TLDR: I would like to stop labeling PRs in [RubyGems](https://github.com/rubygems/rubygems)

**Comment by segiddins** [comment]
I'm obviously biased, but I've been very happy with how we do things over on CocoaPods: https://guides.cocoapods.org/contributing/tickets-handling.html, which also includes not labeling PRs at all.

**Comment by duckinator** [comment]
So, my thoughts on this are basically: I think PR labels could be useful for RubyGems _now_ (if consistently applied), but won't be once we get the number of pull requests under control.

**Comment by segiddins** [comment]
I don't ever look at PR labels -- I look at titles, and CI status, and that's about it (except for notifications)

**Comment by LynnCo** [comment]
Getting the number of PRs (and issues) under control is what I'm going to be working on as soon as I get a handle on how RubyGems is organized, which will be in the next day or so.

**Comment by duckinator** [comment]
Removed "percentage/number of pull requests associated with each label + no label" from the list of metrics to track.

Either of y'all have thoughts on how to present the table more nicely than it currently is, for number of issues per label?

**Comment by segiddins** [comment]
@duckinator have you seen https://github.com/tj/terminal-table ?

**Comment by LynnCo** [comment]
My only thought is that I would want these metrics generated by a webapp daily, howis-rubygems.github.io or something

**Comment by LynnCo** [comment]
Also for said webapp to save the last year or so of stats

**Comment by duckinator** [comment]
@segiddins I meant a better way to organize the information, not how to generate a table (I have screenshots of the PDF version above -- it's easily an entire page just for the list of labels + # of issues for each).

@LynnCo once this is to the point where I release it as a library it should be pretty trivial to make a web app that uses it.

**Comment by indirect** [comment]
I am 👍 on this and you all are doing a great job.

**Comment by duckinator** [comment]
Going with JSON instead of CSV since JSON's actually, you know, doable.

**Comment by duckinator** [comment]
Made it so `how_is --from report.json --report report.pdf` will use the data from ./report.json instead of re-fetching it, so I don't keep running into API limits.

**Comment by duckinator** [comment]
![screenshot 33](https://cloud.githubusercontent.com/assets/39698/15934884/2da04792-2e33-11e6-898e-86ba85b11b03.png)

**Comment by duckinator** [comment]
The average PR/issue age metrics are currently garbage data (hard-coded to `0`), but otherwise I'm pretty happy with where this is going. :3

**Comment by duckinator** [comment]
@LynnCo @segiddins @indirect thoughts?

![screenshot 35](https://cloud.githubusercontent.com/assets/39698/15936422/f8a6ca56-2e38-11e6-9f6c-2a0fa461a0d9.png)

**Comment by duckinator** [comment]
I think, once I'm happy with this, I'm going to add an HTML export and use that as a way to refactor everything, _then_ I'll work on the more complicated metrics.

I want to eventually have something like this, just passed to different renderers:

```ruby
          title "How is #{a.repository}?"

          header "Pull Requests"
          text Report.issue_or_pr_summary(a, "pull", "pull request")

          header "Issues"
          text Report.issue_or_pr_summary(a, "issue", "issue")

          header "Issues Per Label"
          issues_per_label = a.issues_with_label.to_a.sort_by { |(k, v)| v.to_i }.reverse
          table issues_per_label
```

**Comment by duckinator** [comment]
Updated issue to include HTML versions in the task list.

**Comment by duckinator** [comment]
Code: https://github.com/duckinator/how_is/blob/master/lib/how_is/report.rb#L45-L57

Result:
![screenshot 28](https://cloud.githubusercontent.com/assets/39698/15984987/e8f00450-2fab-11e6-8b55-d0a099cabefe.png)

**Comment by duckinator** [comment]
JSON and PDF reports now contain all of the metrics listed in the first groups (the simpler ones).

**Comment by duckinator** [comment]
Separated HTML exports out into #13 and complex metrics out into #12, and everything else in this issue is done, so I'm closing it.
